cpp/src/arrow/util/secure_string_test.cc

```diff
@@ -0,0 +1,478 @@
+  result = AssertSecurelyCleared(long_some_zeros_view);
+  ASSERT_FALSE(result);
+  ASSERT_EQ(std::string(result.message()),
+            "Expected equality of these values:\n"
+            "  area\n"
+            "    Which is: "
+            "\"\\0\\0\\0\\0\\0\\0\\0\\0\\0\\0\\0\\0\\0\\0\\0\\0\\0\\0\\0\\0\\0\\0\\0\\0\\"
+            "0\\0\\0\\0\\0\\0\\0\\0\\0*********\"\n"
+            "  std::string_view(zeros)\n"
+            "    Which is: "
+            "\"\\0\\0\\0\\0\\0\\0\\0\\0\\0\\0\\0\\0\\0\\0\\0\\0\\0\\0\\0\\0\\0\\0\\0\\0\\"
+            "0\\0\\0\\0\\0\\0\\0\\0\\0\\0\\0\\0\\0\\0\\0\\0\\0\\0\"");
+
+  ASSERT_TRUE(AssertSecurelyCleared(long_some_zeros, stars));
+#if !defined(ARROW_VALGRIND) && !defined(ADDRESS_SANITIZER)
+  result = AssertSecurelyCleared(long_some_zeros_view, stars);
+  ASSERT_FALSE(result);
+  ASSERT_EQ(std::string(result.message()),
+            "9 characters of secret leaked into "
+            "\\0\\0\\0\\0\\0\\0\\0\\0\\0\\0\\0\\0\\0\\0\\0\\0\\0\\0\\0\\0\\0\\0\\0\\0\\"
+            "0\\0\\0\\0\\0\\0\\0\\0\\0*********");
+#endif
+
+  // check string with non-zeros and zeros after string length
+  auto some_zeros_back = std::string(no_zeros.length() + 3, '\0');
+  some_zeros_back = no_zeros;
+  memset(some_zeros_back.data() + no_zeros.length() * sizeof(char), '\0', 3 + 1);
+  // string buffer in some_zeros_back can be larger than no_zeros.length() + 3
+  // assert only the area that we can control
+  auto some_zeros_back_view =
+      std::string_view(some_zeros_back.data(), no_zeros.length() + 3);
+  ASSERT_TRUE(AssertSecurelyCleared(some_zeros_back_view, no_zeros.length()));
+}
+
+TEST(TestSecureString, SecureClearString) {
+  // short string
+  {
+    std::string tiny("abc");
+    auto old_area = StringArea(tiny);
+    SecureString::SecureClear(&tiny);
+    ASSERT_TRUE(AssertSecurelyCleared(tiny));
+    ASSERT_TRUE(AssertSecurelyCleared(old_area));
+  }
+
+  // long string
+  {
+    std::string large(1024, 'x');
+    large.resize(512, 'y');
+    auto old_area = StringArea(large);
+    SecureString::SecureClear(&large);
+    ASSERT_TRUE(AssertSecurelyCleared(large));
+    ASSERT_TRUE(AssertSecurelyCleared(old_area));
+  }
+
+  // empty string
+  {
+    // this creates an empty string with some non-zero characters in the string buffer
+    // we test that all those characters are securely cleared
+    std::string empty("abcdef");
+    empty.resize(0);
+    auto old_area = StringArea(empty);
+    SecureString::SecureClear(&empty);
+    ASSERT_TRUE(AssertSecurelyCleared(empty));
+    ASSERT_TRUE(AssertSecurelyCleared(old_area));
+  }
+}
+
+TEST(TestSecureString, Construct) {
+  // move-constructing from a string securely clears that string
+  std::string string("hello world");
+  auto old_string = StringArea(string);
+  SecureString secret_from_string(std::move(string));
+  ASSERT_TRUE(AssertSecurelyCleared(string));
+  ASSERT_TRUE(AssertSecurelyCleared(old_string));
+  ASSERT_FALSE(secret_from_string.empty());
```

Comment:
Perhaps also assert the SecureString contents here?
```suggestion
  ASSERT_FALSE(secret_from_string.empty());
  ASSERT_EQ(secret_from_string.as_view(), "hello world");
```